match2/commons/match_group_display_bracket.lua

```diff
@@ -1,738 +1,745 @@
-    return Bracket.luaGet(frame, args)
+    return html.create('div'):addClass('brkts-main brkts-main-dev')
+        :node(bracketNode)
 end
 
-function Bracket.luaGet(frame, args)
-    _frame = frame
-    
-    local bracketid = args[1]
-    local matches = DisplayHelper.getMatches(bracketid)
-    
-    if (args.matchHeight or '') == '' then
-        for id, match in utils.iter.spairs(matches) do
-            args.matchHeight = WikiSpecific.get_matchHeight(match.match2opponents[1] or {}, match.match2opponents[1] or {}, args.matchHeight)
-        end
-    end
-
-    local has3rd = false
-    for id, match in utils.iter.spairs(matches) do
-        if string.find(match.match2id or '', 'RxMTP') then
-            has3rd = true 
-            break
+Bracket.types.Layout = TypeUtil.struct({
+    height = 'number',
+    lowerMarginTop = 'number',
+    matchHeight = 'number',
+    matchMarginTop = 'number',
+    mid = 'number',
+    showHeader = 'boolean',
+})
+
+--[[
+Computes certain layout properties of nodes in the bracket tree.
+]]
+function Bracket.computeBracketLayout(matchesById, config)
+    -- Map match ids to their upper round matches
+    local upperMatchIds = {}
+    for matchId, match in pairs(matchesById) do
+        for _, x in ipairs(match.bracketData.lowerMatches) do
+            upperMatchIds[x.matchId] = matchId
         end
     end
 
-    _bracketConfig = _getBracketConfig(args)
+    -- Computes the layout of a match and everything to its left.
+    local computeNodeLayout = FnUtil.memoizeY(function(matchId, computeNodeLayout)
+        local match = matchesById[matchId]
+        local lowerLayouts = Array.map(
+            match.bracketData.lowerMatches,
+            function(x) return computeNodeLayout(x.matchId) end
+        )
 
-    local matches, referencedIds = _mapMatches(matches, has3rd)
-    _matches = matches
+        -- Compute partial sums of heights of lower round matches
+        local heightSums = LuaUtils.math.partialSums(
+            Array.map(lowerLayouts, function(layout) return layout.height end)
+        )
 
-    out = html.create('div'):addClass('brkts-main')
-    firsthead = true
-    height = 0
-    for id, match in utils.iter.spairs(matches) do
-        if not referencedIds[match.id] then
-            local extradata, node = unpack(match:buildBracket(firsthead))
-            out:node(node)
-            height = height + math.max(extradata[1], extradata[1] / 2 + match.minDisplayHeight)
-            firsthead = false
+        -- Show a connector line without joints if there is a single lower round 
+        -- match advancing an opponent that is placed near the middle of this match.
+        local singleStraightLine
+        if #lowerLayouts == 1 then
+            local opponentIx = match.bracketData.lowerMatches[1].opponentIx
+            singleStraightLine = 
+                #match.opponents % 2 == 0 
+                    and (opponentIx == #match.opponents / 2 or opponentIx == #match.opponents / 2 + 1)
```

Comment:
Extract this statement (after the `and`) into a boolean so that it is easier to understand